File: custom_components/glt_flow_card/semantic_model.py
"""The validated semantic containment model, mirrored for the Companion.

A rule that exists only in JavaScript is a rule the server does not enforce, so
the graph checks the JSON Schema cannot express - acyclicity, level order,
single parent, depth and breadth bounds - live here as well, and the parity
suite proves the two runtimes agree.
"""
from __future__ import annotations

import json
from collections.abc import Mapping
from pathlib import Path
from typing import Any

_VOCABULARIES = json.loads(
    (Path(__file__).resolve().parent / "schemas" / "vocabularies.json").read_text(encoding="utf-8")
)

SEMANTIC_LEVELS: tuple[str, ...] = tuple(_VOCABULARIES["levels"])
UNITS: dict[str, Any] = dict(_VOCABULARIES["units"])
MEDIA: tuple[str, ...] = tuple(_VOCABULARIES["media"])
DIRECTIONS: tuple[str, ...] = tuple(_VOCABULARIES["directions"])
SEMANTIC_TAGS: tuple[str, ...] = tuple(_VOCABULARIES["semantic_tags"])
BOUNDS: dict[str, int] = dict(_VOCABULARIES["bounds"])

_LEVEL_INDEX = {level: index for index, level in enumerate(SEMANTIC_LEVELS)}


def _error(code: str, path: str, detail: Mapping[str, Any] | None = None) -> dict[str, Any]:
    error = {"code": code, "path": path}
    if detail is not None:
        error["detail"] = dict(detail)
    return error


def _nodes(model: Any) -> list[Any]:
    nodes = model.get("nodes") if isinstance(model, Mapping) else None
    return list(nodes) if isinstance(nodes, list) else []


def validate_semantic_model(model: Any) -> list[dict[str, Any]]:
    """Return every problem with a semantic model, empty when acceptable."""
    nodes = _nodes(model)
    errors: list[dict[str, Any]] = []

    if len(nodes) > BOUNDS["max_nodes"]:
        return [_error("semantic_model_too_large", "/semantic_model/nodes",
                       {"nodes": len(nodes), "max": BOUNDS["max_nodes"]})]

    by_id: dict[str, tuple[Mapping[str, Any], str]] = {}
    children: dict[str, int] = {}
    for index, node in enumerate(nodes):
        path = f"/semantic_model/nodes/{index}"
        if not isinstance(node, Mapping) or not isinstance(node.get("id"), str) or not node["id"]:
            errors.append(_error("semantic_node_id_missing", path))
            continue
        if node["id"] in by_id:
            errors.append(_error("semantic_node_duplicate_id", path, {"id": node["id"]}))
            continue
        by_id[node["id"]] = (node, path)

    for node, path in by_id.values():
        if node.get("level") not in _LEVEL_INDEX:
            errors.append(_error("semantic_level_unknown", path, {"level": node.get("level")}))
        for field, vocabulary in (("unit", UNITS), ("medium", MEDIA), ("direction", DIRECTIONS)):
            value = node.get(field)
            if value is None:
                continue
            known = value in vocabulary
            if not known:
                errors.append(_error(f"semantic_{field}_unknown", f"{path}/{field}", {"value": value}))
        tags = node.get("semantic_tags")
        if isinstance(tags, list):
            for tag_index, tag in enumerate(tags):
                if tag not in SEMANTIC_TAGS:
                    errors.append(_error("semantic_tag_unknown",
                                         f"{path}/semantic_tags/{tag_index}", {"tag": tag}))

        parent = node.get("parent")
        if parent is None:
            continue
        target = by_id.get(parent)
        if target is None:
            errors.append(_error("semantic_parent_missing", f"{path}/parent", {"parent": parent}))
            continue
        children[parent] = children.get(parent, 0) + 1
        parent_level = _LEVEL_INDEX.get(target[0].get("level"))
        node_level = _LEVEL_INDEX.get(node.get("level"))
        if parent_level is not None and node_level is not None:
            if node_level < parent_level:
                errors.append(_error("semantic_level_inverted", f"{path}/parent",
                                     {"level": node.get("level"), "parent_level": target[0].get("level")}))
            elif target[0].get("level") == "datapoint":
                errors.append(_error("semantic_datapoint_has_child", f"{path}/parent", {"parent": parent}))

    for parent, count in children.items():
        if count > BOUNDS["max_children"]:
            errors.append(_error("semantic_children_exceeded",
                                 by_id.get(parent, ({}, "/semantic_model/nodes"))[1],
                                 {"children": count, "max": BOUNDS["max_children"]}))

    for node_id, (node, path) in by_id.items():
        seen = {node_id}
        current = node.get("parent")
        depth = 1
        while current is not None:
            if current in seen:
                errors.append(_error("semantic_cycle", f"{path}/parent", {"closes_at": current}))
                break
            seen.add(current)
            following = by_id.get(current)
            if following is None:
                break
            depth += 1
            if depth > BOUNDS["max_depth"]:
                errors.append(_error("semantic_depth_exceeded", path,
                                     {"depth": depth, "max": BOUNDS["max_depth"]}))
                break
            current = following[0].get("parent")

    return errors


def semantic_path(model: Any, node_id: str) -> list[str]:
    """Derive a node's containment path, root first. Never stored."""
    by_id = {node["id"]: node for node in _nodes(model)
             if isinstance(node, Mapping) and isinstance(node.get("id"), str)}
    path: list[str] = []
    seen: set[str] = set()
    current: Any = node_id
    while current is not None and current in by_id and current not in seen:
        seen.add(current)
        path.insert(0, current)
        current = by_id[current].get("parent")
    return path


def same_dimension(left: str, right: str) -> bool:
    """Whether two units can be compared or summed at all."""
    a = UNITS.get(left)
    b = UNITS.get(right)
    return bool(a and b and a["dimension"] == b["dimension"])

File: src/v100/semantic-model.mjs
/* The validated semantic containment model.
 *
 * A hierarchy that accepts a cycle is a hierarchy nothing can safely walk:
 * navigation loops, permission inheritance never terminates, and a roll-up sums
 * a node into itself. Reference checking alone cannot catch that — it proves a
 * parent exists, not that the graph has a bottom — so the graph rules live here
 * and run in both runtimes.
 *
 * The vocabularies come from the schema that enforces them, so the module and
 * the contract cannot hold two different opinions about what a unit is.
 */
import { contractVocabularies } from "./generated/project-validators.mjs";

/** Containment order, outermost first. */
export const SEMANTIC_LEVELS = Object.freeze([...contractVocabularies.levels]);

/** Unit symbol -> its dimension. Two units of different dimensions never mix. */
export const UNITS = Object.freeze(contractVocabularies.units);
export const MEDIA = Object.freeze([...contractVocabularies.media]);
export const DIRECTIONS = Object.freeze([...contractVocabularies.directions]);
export const SEMANTIC_TAGS = Object.freeze([...contractVocabularies.semantic_tags]);

/** Structural bounds. A hostile document must not make validation unbounded. */
export const BOUNDS = Object.freeze({ ...contractVocabularies.bounds });

const LEVEL_INDEX = Object.freeze(Object.fromEntries(
  SEMANTIC_LEVELS.map((level, index) => [level, index]),
));

function error(code, path, detail) {
  return detail === undefined ? { code, path } : { code, path, detail };
}

function nodesOf(model) {
  const nodes = model?.nodes;
  return Array.isArray(nodes) ? nodes : [];
}

/**
 * Validate one semantic model and return every problem found.
 *
 * Errors accumulate rather than short-circuiting: an engineer fixing a model
 * wants the whole list, not one problem per round trip. Every error carries a
 * stable code and the path of the node that caused it.
 */
export function validateSemanticModel(model) {
  const nodes = nodesOf(model);
  const errors = [];

  if (nodes.length > BOUNDS.max_nodes) {
    errors.push(error("semantic_model_too_large", "/semantic_model/nodes",
      { nodes: nodes.length, max: BOUNDS.max_nodes }));
    return errors;
  }

  const byId = new Map();
  const children = new Map();
  nodes.forEach((node, index) => {
    const path = `/semantic_model/nodes/${index}`;
    const id = node?.id;
    if (typeof id !== "string" || id.length === 0) {
      errors.push(error("semantic_node_id_missing", path));
      return;
    }
    if (byId.has(id)) {
      // Two nodes with one id is how a second parent gets in.
      errors.push(error("semantic_node_duplicate_id", path, { id }));
      return;
    }
    byId.set(id, { node, path });
  });

  for (const [id, { node, path }] of byId) {
    if (!Object.hasOwn(LEVEL_INDEX, node.level)) {
      errors.push(error("semantic_level_unknown", path, { level: node.level ?? null }));
    }
    for (const [field, vocabulary] of [
      ["unit", UNITS], ["medium", MEDIA], ["direction", DIRECTIONS],
    ]) {
      const value = node[field];
      if (value === undefined || value === null) continue;
      const known = Array.isArray(vocabulary) ? vocabulary.includes(value) : Object.hasOwn(vocabulary, value);
      if (!known) errors.push(error(`semantic_${field}_unknown`, `${path}/${field}`, { value }));
    }
    if (Array.isArray(node.semantic_tags)) {
      node.semantic_tags.forEach((tag, tagIndex) => {
        if (!SEMANTIC_TAGS.includes(tag)) {
          errors.push(error("semantic_tag_unknown", `${path}/semantic_tags/${tagIndex}`, { tag }));
        }
      });
    }

    const parent = node.parent;
    if (parent === undefined || parent === null) continue;
    const target = byId.get(parent);
    if (target === undefined) {
      errors.push(error("semantic_parent_missing", `${path}/parent`, { parent }));
      continue;
    }
    children.set(parent, (children.get(parent) ?? 0) + 1);
    // A child may sit at its parent's level or deeper — a subsystem inside a
    // subsystem is ordinary plant — but never above it, and nothing lives
    // inside a datapoint.
    const parentLevel = LEVEL_INDEX[target.node.level];
    const nodeLevel = LEVEL_INDEX[node.level];
    if (parentLevel !== undefined && nodeLevel !== undefined) {
      if (nodeLevel < parentLevel) {
        errors.push(error("semantic_level_inverted", `${path}/parent`,
          { level: node.level, parent_level: target.node.level }));
      } else if (target.node.level === "datapoint") {
        errors.push(error("semantic_datapoint_has_child", `${path}/parent`, { parent }));
      }
    }
    void id;
  }

  for (const [parent, count] of children) {
    if (count > BOUNDS.max_children) {
      errors.push(error("semantic_children_exceeded",
        `${byId.get(parent)?.path ?? "/semantic_model/nodes"}`,
        { children: count, max: BOUNDS.max_children }));
    }
  }

  for (const [id, { path }] of byId) {
    const seen = new Set([id]);
    let current = byId.get(id).node.parent;
    let depth = 1;
    while (current !== undefined && current !== null) {
      if (seen.has(current)) {
        errors.push(error("semantic_cycle", `${path}/parent`, { closes_at: current }));
        break;
      }
      seen.add(current);
      const next = byId.get(current);
      if (next === undefined) break;
      depth += 1;
      if (depth > BOUNDS.max_depth) {
        errors.push(error("semantic_depth_exceeded", path, { depth, max: BOUNDS.max_depth }));
        break;
      }
      current = next.node.parent;
    }
  }

  return errors;
}

/**
 * Derive a node's containment path, root first.
 *
 * The path is never stored. A stored path is a second source of truth that
 * starts agreeing with its parents and stops without telling anyone.
 */
export function semanticPath(model, nodeId) {
  const byId = new Map(nodesOf(model).map((node) => [node?.id, node]));
  const path = [];
  const seen = new Set();
  let current = nodeId;
  while (current !== undefined && current !== null && byId.has(current) && !seen.has(current)) {
    seen.add(current);
    path.unshift(current);
    current = byId.get(current).parent;
  }
  return path;
}

/** Whether two units can be compared or summed at all. */
export function sameDimension(left, right) {
  const a = UNITS[left];
  const b = UNITS[right];
  return Boolean(a && b && a.dimension === b.dimension);
}
